provider/src/main/kotlin/org/gradle/kotlin/dsl/support/KotlinCompiler.kt

```diff
@@ -204,44 +205,60 @@ fun kotlinCoreEnvironmentFor(configuration: CompilerConfiguration, rootDisposabl
 
 
 internal
-fun messageCollectorFor(log: Logger, pathTranslation: (String) -> String = { it }): MessageCollector =
+fun messageCollectorFor(log: Logger, pathTranslation: (String) -> String = { it }): LoggingMessageCollector =
+    LoggingMessageCollector(log, pathTranslation)
 
-    object : MessageCollector {
 
-        var errors = 0
+internal
+data class ScriptCompilationError(val message: String, val lineNumber: Int?)
 
-        override fun hasErrors() = errors > 0
 
-        override fun clear() { errors = 0 }
+internal
+data class ScriptCompilationException(val errors: List<ScriptCompilationError>) : RuntimeException() {
+    override val message: String?
+        get() = errors.first().message
+}
 
-        override fun report(severity: CompilerMessageSeverity, message: String, location: CompilerMessageLocation?) {
 
-            fun msg() =
-                location?.run {
-                    path.let(pathTranslation).let { path ->
-                        when {
-                            line >= 0 && column >= 0 -> compilerMessageFor(path, line, column, message)
-                            else -> "$path: $message"
-                        }
-                    }
-                } ?: message
+internal
+class LoggingMessageCollector(
+    private val log: Logger,
+    private val pathTranslation: (String) -> String) : MessageCollector {
+
+    val errors = arrayListOf<ScriptCompilationError>()
 
-            fun taggedMsg() =
-                "${severity.presentableName[0]}: ${msg()}"
+    override fun hasErrors() = errors.isNotEmpty()
 
-            when (severity) {
-                CompilerMessageSeverity.ERROR, CompilerMessageSeverity.EXCEPTION -> {
-                    errors++
-                    log.error { taggedMsg() }
+    override fun clear() { errors.clear() }
+
+    override fun report(severity: CompilerMessageSeverity, message: String, location: CompilerMessageLocation?) {
+
+        fun msg() =
+            location?.run {
+                path.let(pathTranslation).let { path ->
+                    when {
+                        line >= 0 && column >= 0 -> compilerMessageFor(path, line, column, message)
+                        else -> "$path: $message"
+                    }
                 }
-                in CompilerMessageSeverity.VERBOSE -> log.trace { msg() }
-                CompilerMessageSeverity.STRONG_WARNING -> log.info { taggedMsg() }
-                CompilerMessageSeverity.WARNING -> log.info { taggedMsg() }
-                CompilerMessageSeverity.INFO -> log.info { msg() }
-                else -> log.debug { taggedMsg() }
+            } ?: message
+
+        fun taggedMsg() =
+            "${severity.presentableName[0]}: ${msg()}"
+
+        when (severity) {
+            CompilerMessageSeverity.ERROR, CompilerMessageSeverity.EXCEPTION -> {
+                errors += ScriptCompilationError(message, location?.line)
```

Comment:
Is `report` guaranteed to only be called from one thread or could it be called by two threads concurrently?
If it could be called by two threads concurrently you could get `ConcurrentModificationException`

Reply:
Good question.

@ligee do you know if Kotlin compiler's `MessageCollector` could be invoked concurrently?

Reply:
Now, generally - not, but we cannot guarantee that it remains like this in the future.